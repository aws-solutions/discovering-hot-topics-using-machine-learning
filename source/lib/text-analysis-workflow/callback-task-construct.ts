/**********************************************************************************************************************
* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
* SPDX-License-Identifier: Apache-2.0
 *********************************************************************************************************************/

import { SqsToLambda } from '@aws-solutions-constructs/aws-sqs-lambda';
import * as defaults from '@aws-solutions-constructs/core';
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { SqsEventSourceProps } from 'aws-cdk-lib/aws-lambda-event-sources';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as sfn from 'aws-cdk-lib/aws-stepfunctions';
import * as task from 'aws-cdk-lib/aws-stepfunctions-tasks';
import { Construct } from 'constructs';

export interface StepFuncCallbackTaskProps {
    /**
     * Existing instance of Lambda Function object, providing both this and `lambdaFunctionProps` will cause an error.
     *
     * @default - None
     */
    readonly existingLambdaObj?: lambda.Function,
    /**
     * User provided props to override the default props for the Lambda function.
     *
     * @default - Default properties are used.
     */
    readonly lambdaFunctionProps?: lambda.FunctionProps,
    /**
     * Existing instance of SQS queue object, Providing both this and queueProps will cause an error.
     *
     * @default - Default props are used
     */
    readonly existingQueueObj?: sqs.Queue;
    /**
     * Optional user-provided props to override the default props for the SQS queue.
     *
     * @default - Default props are used
     */
    readonly queueProps?: sqs.QueueProps;
    /**
     * Optional user provided properties for the dead letter queue
     *
     * @default - Default props are used
     */
    readonly deadLetterQueueProps?: sqs.QueueProps;
    /**
     * Whether to deploy a secondary queue to be used as a dead letter queue.
     *
     * @default - true.
     */
    readonly deployDeadLetterQueue?: boolean;
    /**
     * The number of times a message can be unsuccessfully dequeued before being moved to the dead-letter queue.
     *
     * @default - required field if deployDeadLetterQueue=true.
     */
    readonly maxReceiveCount?: number;
    /**
     * Optional user provided properties for the dead letter queue
     *
     * @default - Default props are used
     */
    readonly sqsEventSourceProps?: SqsEventSourceProps
    /**
     * Optional user provided properties to create an SQS based stepfunction task
     *
     * @default - None
     */
    readonly sqsSendMessageProps?: Partial<task.SqsSendMessageProps>;
    /**
     * Optional user provided state machine that the lambda function can sendTaskResponse. This role policy of
     * the lambda function will be restricted to be able to send the response token only to the provided state
     * machine. If the parameter is not supplied, the lambda role policy for sending the task token response
     * will contain '*' in the resources section
     *
     * @default - None
     */
    readonly stateMachine?: sfn.StateMachine;
}

export class StepFuncCallbackTask extends sfn.StateMachineFragment {
    public readonly sqsQueue: sqs.Queue;
    public readonly deadLetterQueue: sqs.DeadLetterQueue | undefined;
    public readonly lambda: lambda.Function;
    public readonly startState: sfn.State;
    public readonly endStates: sfn.INextable[];

    constructor(scope: Construct, id: string, props: StepFuncCallbackTaskProps) {
        super(scope, id);

        const _sqsLambda = new SqsToLambda(this, 'Task', {
            lambdaFunctionProps: props.lambdaFunctionProps,
            existingLambdaObj: props.existingLambdaObj,
            queueProps: props.queueProps,
            existingQueueObj: props.existingQueueObj,
            deadLetterQueueProps: props.deadLetterQueueProps,
            deployDeadLetterQueue: props.deployDeadLetterQueue,
            maxReceiveCount: props.maxReceiveCount,
            sqsEventSourceProps: props.sqsEventSourceProps
        });

        const _defaultSqsSendMessageProps: task.SqsSendMessageProps = {
            queue: _sqsLambda.sqsQueue,
            messageBody: sfn.TaskInput.fromObject({
                "input": sfn.JsonPath.stringAt('$'),
                "taskToken": sfn.JsonPath.taskToken
            }),
            integrationPattern: sfn.IntegrationPattern.WAIT_FOR_TASK_TOKEN,
            outputPath: "$",
            heartbeatTimeout: sfn.Timeout.duration(cdk.Duration.minutes(720)),
            taskTimeout: sfn.Timeout.duration(cdk.Duration.minutes(1440))
        };

        const _sqsSendMessageProps = props.sqsSendMessageProps !== undefined ?
            defaults.overrideProps(_defaultSqsSendMessageProps, props.sqsSendMessageProps as task.SqsSendMessageProps):
            _defaultSqsSendMessageProps;

        const _queueTask: task.SqsSendMessage = new task.SqsSendMessage(this, `${id}SendMessage`, _sqsSendMessageProps);


        const _failedState = new sfn.Fail(this, `${id}TaskFailed`, {
            cause: sfn.JsonPath.stringAt('$.cause')
        });
        _queueTask.addCatch(_failedState);

        _queueTask.addRetry({
            backoffRate: 2,
            maxAttempts: 6,
            interval: cdk.Duration.seconds(3)
        });

        this.startState = _queueTask;
        this.endStates = [ _queueTask ]
        this.sqsQueue = _sqsLambda.sqsQueue;
        this.lambda = _sqsLambda.lambdaFunction;
        this.deadLetterQueue = _sqsLambda.deadLetterQueue;

        if (props.stateMachine !== undefined) {
            props.stateMachine.grantTaskResponse(this.lambda.role!);
        } else {
            const _lambdaStateMatchineTaskPolicy = new iam.Policy(this, 'LambdaStateMachineTask', {
                statements: [ new iam.PolicyStatement({
                        effect: iam.Effect.ALLOW,
                        actions: [ 'states:SendTaskSuccess', 'states:sendTaskFailure', 'states:SendTaskHeartbeat' ],
                        resources: [ '*' ]
                    })
                ]
            });

            _lambdaStateMatchineTaskPolicy.attachToRole(_sqsLambda.lambdaFunction.role!);
            (_lambdaStateMatchineTaskPolicy.node.defaultChild as iam.CfnPolicy).addMetadata('cfn_nag', {
                rules_to_suppress: [{
                  id: 'W12',
                  reason: 'State machine resource not available, hence defaulting to "*"'
                }]
            });
        }
    }
}