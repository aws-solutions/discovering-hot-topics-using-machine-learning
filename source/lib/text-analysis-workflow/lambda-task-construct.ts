/**********************************************************************************************************************
* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
* SPDX-License-Identifier: Apache-2.0
 *********************************************************************************************************************/


import * as defaults from '@aws-solutions-constructs/core';
import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as sfn from 'aws-cdk-lib/aws-stepfunctions';
import * as tasks from 'aws-cdk-lib/aws-stepfunctions-tasks';
import { Construct } from 'constructs';

export interface StepFuncLambdaTaskProps {
    readonly taskName: string,
    readonly lambdaFunctionProps: lambda.FunctionProps,
    readonly inputPath?: string | any,
    readonly outputPath?: string | any
}

export class StepFuncLambdaTask extends Construct {

    private executeTask: tasks.LambdaInvoke;

    private lambdaFn: lambda.Function;

    constructor (scope: Construct, id: string, props: StepFuncLambdaTaskProps) {
        super(scope, id);

        this.lambdaFn = defaults.buildLambdaFunction(this, {
            lambdaFunctionProps: props.lambdaFunctionProps,
        });

        this.executeTask = new tasks.LambdaInvoke(this, props.taskName, {
            lambdaFunction: this.lambdaFn,
            inputPath: props.inputPath,
            outputPath: props.outputPath,
            heartbeatTimeout: sfn.Timeout.duration(cdk.Duration.minutes(15)),
            retryOnServiceExceptions: true,
        });

        const _failedState = new sfn.Fail(this, `${id}TaskFailed`, {
            cause: sfn.JsonPath.stringAt('$.cause')
        });

        this.executeTask.addCatch(_failedState);
    }

    public get stepFunctionTask(): tasks.LambdaInvoke {
        return this.executeTask;
    }

    public get lambdaFunction(): lambda.Function {
        return this.lambdaFn;
    }
}
