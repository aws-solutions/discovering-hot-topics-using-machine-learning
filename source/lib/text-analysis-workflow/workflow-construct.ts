/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { LambdaToStepfunctions } from '@aws-solutions-constructs/aws-lambda-stepfunctions';
import * as cdk from 'aws-cdk-lib';
import { CfnPolicy, Effect, Policy, PolicyStatement, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Function, IFunction } from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as sfn from 'aws-cdk-lib/aws-stepfunctions';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';

export interface WorkflowProps {
    readonly stateMachineType?: sfn.StateMachineType;
    readonly chain: sfn.Chain;
    readonly lambdaFunc?: IFunction;
}

export class Workflow extends Construct {
    private _stMachine: sfn.StateMachine;

    constructor(scope: Construct, id: string, props: WorkflowProps) {
        super(scope, id);

        if (props.lambdaFunc != null && props.lambdaFunc != undefined) {
            const lambdaStepFunction = new LambdaToStepfunctions(this, 'WorkflowEngine', {
                existingLambdaObj: props.lambdaFunc as Function,
                stateMachineProps: {
                    definitionBody: sfn.DefinitionBody.fromChainable(props.chain),
                    tracingEnabled: true,
                    ...(props.stateMachineType === sfn.StateMachineType.EXPRESS && {
                        stateMachineType: props.stateMachineType,
                        logs: {
                            level: sfn.LogLevel.ERROR,
                            includeExecutionData: false,
                            destination: new logs.LogGroup(this, 'TextAnalysisWF', {
                                logGroupName: `/aws/vendedlogs/states/${cdk.Stack.of(this).getLogicalId}`.slice(0, 255)
                            })
                        }
                    })
                }
            });
            this._stMachine = lambdaStepFunction.stateMachine;
            const role = this._stMachine.node.findChild('Role') as Role;
            const cfnDefaultPolicy = role.node.findChild('DefaultPolicy').node.defaultChild as CfnPolicy;

            cfnDefaultPolicy.addMetadata('cfn_nag', {
                rules_to_suppress: [
                    {
                        id: 'W76',
                        reason:
                            'The policy adds cloudwatch alarms and allows step function to invoke lambda tasks. Suppressing the ' +
                            'SPCM violation as this policy is required to monitor as well as invoke specific tasks'
                    },
                    {
                        id: 'W12',
                        reason: "The 'LogDelivery' actions do not support resource-level authorizations"
                    }
                ]
            });
        } else {
            const _stateMachineRole = new Role(this, 'StateMachineRole', {
                assumedBy: new ServicePrincipal(`states.${cdk.Aws.REGION}.amazonaws.com`)
            });

            const stateMachineLogPolicy = new Policy(this, 'StateMachineLogPolicy', {
                statements: [
                    new PolicyStatement({
                        actions: [
                            'logs:CreateLogDelivery',
                            'logs:GetLogDelivery',
                            'logs:UpdateLogDelivery',
                            'logs:DeleteLogDelivery',
                            'logs:ListLogDeliveries',
                            'logs:PutResourcePolicy',
                            'logs:DescribeResourcePolicies',
                            'logs:DescribeLogGroups'
                        ],
                        resources: [`arn:${cdk.Aws.PARTITION}:logs:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:*`],
                        effect: Effect.ALLOW
                    })
                ]
            });

            _stateMachineRole.attachInlinePolicy(stateMachineLogPolicy);

            this._stMachine = new sfn.StateMachine(this, 'WorkflowEngine', {
                definitionBody: sfn.DefinitionBody.fromChainable(props.chain),
                tracingEnabled: true,
                ...(props.stateMachineType === sfn.StateMachineType.EXPRESS && {
                    stateMachineType: props.stateMachineType,
                    role: _stateMachineRole,
                    logs: {
                        level: sfn.LogLevel.ERROR,
                        includeExecutionData: false,
                        destination: new logs.LogGroup(this, 'TextAnalysisWF', {
                            logGroupName: `/aws/vendedlogs/states/${cdk.Stack.of(this).getLogicalId}`.slice(0, 255)
                        })
                    }
                })
            });

            this._stMachine.node.addDependency(_stateMachineRole);
        }
        NagSuppressions.addResourceSuppressions(
            this._stMachine, [
            {
                id: 'AwsSolutions-SF1', reason: "Information required for troubleshooting is logged by state function and lambda functions",
            }
        ],
            true
        );
    }

    public get stateMachine(): sfn.StateMachine {
        return this._stMachine;
    }
}
