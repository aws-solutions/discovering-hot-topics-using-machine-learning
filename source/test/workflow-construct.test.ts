/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Template } from 'aws-cdk-lib/assertions';
import { Chain, StateMachineType, Succeed, Wait, WaitTime } from 'aws-cdk-lib/aws-stepfunctions';
import * as cdk from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { Workflow } from '../lib/text-analysis-workflow/workflow-construct';


test('test standard workflow stack', () => {
    const stack = new cdk.Stack();
    new Workflow (stack, 'WorkflowStack', {
        chain: Chain.start(new Wait(stack, 'WaitState', {
                        time: WaitTime.duration(Duration.minutes(1)),
                    }))
                    .next(new Succeed(stack, 'Success'))
    });
    });

test('test express workflow stack', () => {
    const stack = new cdk.Stack();
    new Workflow (stack, 'WorkflowStack', {
        stateMachineType: StateMachineType.EXPRESS,
        chain: Chain.start(new Wait(stack, 'WaitState', {
                        time: WaitTime.duration(Duration.minutes(1)),
                    }))
                    .next(new Succeed(stack, 'Success'))
    });
    });