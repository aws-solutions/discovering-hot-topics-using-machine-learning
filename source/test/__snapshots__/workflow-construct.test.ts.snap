// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test express workflow stack 1`] = `
{
  "Resources": {
    "WorkflowStackStateMachineLogPolicyB222B557": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogDelivery",
                "logs:GetLogDelivery",
                "logs:UpdateLogDelivery",
                "logs:DeleteLogDelivery",
                "logs:ListLogDeliveries",
                "logs:PutResourcePolicy",
                "logs:DescribeResourcePolicies",
                "logs:DescribeLogGroups",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WorkflowStackStateMachineLogPolicyB222B557",
        "Roles": [
          {
            "Ref": "WorkflowStackStateMachineRole56AD52FB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WorkflowStackStateMachineRole56AD52FB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::Join": [
                    "",
                    [
                      "states.",
                      {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "WorkflowStackStateMachineRoleDefaultPolicy25244020": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogDelivery",
                "logs:GetLogDelivery",
                "logs:UpdateLogDelivery",
                "logs:DeleteLogDelivery",
                "logs:ListLogDeliveries",
                "logs:PutResourcePolicy",
                "logs:DescribeResourcePolicies",
                "logs:DescribeLogGroups",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WorkflowStackStateMachineRoleDefaultPolicy25244020",
        "Roles": [
          {
            "Ref": "WorkflowStackStateMachineRole56AD52FB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WorkflowStackTextAnalysisWFB1867CEA": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "/aws/vendedlogs/states/getLogicalId(element) {
        try {
            jsiiDeprecationWarnings._aws_cdk_core_CfnElement(element);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
  ",
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "WorkflowStackWorkflowEngine48A8BD9B": {
      "DependsOn": [
        "WorkflowStackStateMachineRoleDefaultPolicy25244020",
        "WorkflowStackStateMachineRole56AD52FB",
      ],
      "Properties": {
        "DefinitionString": "{"StartAt":"WaitState","States":{"WaitState":{"Type":"Wait","Seconds":60,"Next":"Success"},"Success":{"Type":"Succeed"}}}",
        "LoggingConfiguration": {
          "Destinations": [
            {
              "CloudWatchLogsLogGroup": {
                "LogGroupArn": {
                  "Fn::GetAtt": [
                    "WorkflowStackTextAnalysisWFB1867CEA",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "IncludeExecutionData": false,
          "Level": "ERROR",
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "WorkflowStackStateMachineRole56AD52FB",
            "Arn",
          ],
        },
        "StateMachineType": "EXPRESS",
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
  },
}
`;

exports[`test standard workflow stack 1`] = `
{
  "Mappings": {
    "ServiceprincipalMap": {
      "af-south-1": {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-southeast-1": {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ca-central-1": {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-north-1": {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "me-south-1": {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-west-1": {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Resources": {
    "WorkflowStackStateMachineLogPolicyB222B557": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogDelivery",
                "logs:GetLogDelivery",
                "logs:UpdateLogDelivery",
                "logs:DeleteLogDelivery",
                "logs:ListLogDeliveries",
                "logs:PutResourcePolicy",
                "logs:DescribeResourcePolicies",
                "logs:DescribeLogGroups",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WorkflowStackStateMachineLogPolicyB222B557",
        "Roles": [
          {
            "Ref": "WorkflowStackStateMachineRole56AD52FB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WorkflowStackStateMachineRole56AD52FB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::Join": [
                    "",
                    [
                      "states.",
                      {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "WorkflowStackWorkflowEngine48A8BD9B": {
      "DependsOn": [
        "WorkflowStackStateMachineRole56AD52FB",
        "WorkflowStackWorkflowEngineRoleF56F2281",
      ],
      "Properties": {
        "DefinitionString": "{"StartAt":"WaitState","States":{"WaitState":{"Type":"Wait","Seconds":60,"Next":"Success"},"Success":{"Type":"Succeed"}}}",
        "RoleArn": {
          "Fn::GetAtt": [
            "WorkflowStackWorkflowEngineRoleF56F2281",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "WorkflowStackWorkflowEngineRoleF56F2281": {
      "DependsOn": [
        "WorkflowStackStateMachineRole56AD52FB",
      ],
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::FindInMap": [
                    "ServiceprincipalMap",
                    {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
